inc/functions_search.php

```diff
@@ -150,6 +150,71 @@ function get_unsearchable_forums($pid=0, $first=1)
 	return $unsearchable;
 }
 
+/**
+ * Build query condition for threads/posts the user is allowed to see.
+ * Will return for example:
+ *  - visibile = 1 - for normal users
+ *  - visible >=1 - for admins & super mods
+ *  - (visibile = 1 OR (visibile = ? AND fid IN ...)) - for forum moderators
+ * 
+ * @param string $table_alias The alias of the table eg t to use t.visible instead of visible
+ * @return string the query condition 
+ */
+function get_visible_where($table_alias = null)
+{
+	global $db, $mybb;
+
+	$aliasdot = '';
+	if(!empty($table_alias))
+	{
+		$aliasdot = $table_alias.'.';
+	}
+
+	$query = $db->simple_select("moderators", "fid, canviewunapprove, canviewdeleted", "(id='{$mybb->user['uid']}' AND isgroup='0') OR (id='{$mybb->user['usergroup']}' AND isgroup='1')");
```

Comment:
`is_moderator()` or `get_moderator_permissions()` should be used here since records in table `moderators` don't have permission settings for inherited forums (child forums) if no explicit moderator assignment exists.

Note: any of the two functions works on forum level, actually one is calling the other, and thus if a search without forum specified (meaning to search in all forums), we may have to find current user's relevant permissions in all forums.

Reply:
Hmm so I was using the code that was already present and just extracted to a function. I looked at `is_moderator` initially instead of this but really what I wanted was this code that already existed.

I suppose to use `is_moderator` here we would need to loop over all forums to get this list of fids and visibility pairings. What do you think?

Reply:
> I suppose to use `is_moderator` here we would need to loop over all forums to get this list of fids and visibility pairings. What do you think?

I do think so.

And then we might have to strip some code from [`get_moderator_permissions()`](https://github.com/mybb/mybb/blob/mybb_1822/inc/functions.php#L1621-L1720) to get a list/array of forums a user moderates in, probably as an array `$moderatingfids`.

And further we may do something here (may not work well):
```
$permsql = '';
// Check group permissions if we can't view threads not started by us
$onlyusfids = array();
if($group_permissions = forum_permissions())
{
	foreach($group_permissions as $fid => $forum_permissions)
	{
		if(isset($forum_permissions['canonlyviewownthreads']) && $forum_permissions['canonlyviewownthreads'] == 1)
		{
			$onlyusfids[] = $fid;
		}
	}
}

if(!empty($onlyusfids))
{
	$permsql .= "(t.fid IN(".implode(',', $onlyusfids).") AND t.uid='{$mybb->user['uid']}') OR t.fid NOT IN(".implode(',', $onlyusfids).")";
}

$moderating_and_onlyus_fids = array_intersect($onlyusfids, $moderatingfids);
$unapprove_forums = $deleted_forums = array();
foreach($moderating_and_onlyus_fid) as $fid)
{
	if(is_moderator($fid, 'canviewunapprove'))
	{
		$unapprove_forums[] = $fid;
	}
	if(is_moderator($fid, 'canviewdeleted'))
	{
		$deleted_forums[] = $fid;
	}
}
if(!empty($unapprove_forums))
{
	$permsql .= " OR (t.uid!='{$mybb->user['uid']}' AND t.visible = 0 AND t.fid IN(".implode(',', $unapprove_forums)."))";
}
if(!empty($deleted_forums))
{
	$permsql .= " OR (t.uid!='{$mybb->user['uid']}' AND t.visible = -1 AND t.fid IN(".implode(',', $deleted_forums)."))";
}

$moderating_not_onlyus_fids = array_diff($moderatingfids, $moderating_and_onlyus_fids);
$unapprove_forums = $deleted_forums = array();

// original code
...

return 'AND ({$permsql})';

```

Reply:
Or perhaps the function may return an associate array with 't' and 'p' if using a parameter to control the return becomes troublesome.